### PR TITLE
[wasm] [DebuggerProxy] Capture close of connection without proper handshake

### PR DESCRIPTION
**Pull request by kjpou1**
```

06:01:29 connecting to ws://localhost:9300/launch-chrome-and-connect
06:01:29 got exception System.AggregateException: One or more errors occurred. (The remote party closed the WebSocket connection without completing the close handshake.) ---> System.Net.WebSockets.WebSocketException: The remote party closed the WebSocket connection without completing the close handshake. ---> System.Net.WebSockets.WebSocketException: The remote party closed the WebSocket connection without completing the close handshake.
06:01:29    at System.Net.WebSockets.ManagedWebSocket.ThrowIfEOFUnexpected(Boolean throwOnPrematureClosure)
06:01:29    at System.Net.WebSockets.ManagedWebSocket.EnsureBufferContainsAsync(Int32 minimumRequiredBytes, Boolean throwOnPrematureClosure)
06:01:29    at System.Net.WebSockets.ManagedWebSocket.ReceiveAsyncPrivate[TWebSocketReceiveResultGetter,TWebSocketReceiveResult](Memory`1 payloadBuffer, CancellationToken cancellationToken, TWebSocketReceiveResultGetter resultGetter)
06:01:29    --- End of inner exception stack trace ---
06:01:29    at System.Net.WebSockets.ManagedWebSocket.ReceiveAsyncPrivate[TWebSocketReceiveResultGetter,TWebSocketReceiveResult](Memory`1 payloadBuffer, CancellationToken cancellationToken, TWebSocketReceiveResultGetter resultGetter)
06:01:29    at WsProxy.WsProxy.ReadOne(WebSocket socket, CancellationToken token) in /mnt/jenkins/workspace/test-mono-pull-request-wasm/sdks/wasm/Mono.WebAssembly.DebuggerProxy/WsProxy.cs:line 122
06:01:29    --- End of inner exception stack trace ---
06:01:29    at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
06:01:29    at WsProxy.WsProxy.Run(Uri browserUri, WebSocket ideSocket) in /mnt/jenkins/workspace/test-mono-pull-request-wasm/sdks/wasm/Mono.WebAssembly.DebuggerProxy/WsProxy.cs:line 286
06:01:29 ---> (Inner Exception #0) System.Net.WebSockets.WebSocketException (0x80004005): The remote party closed the WebSocket connection without completing the close handshake. ---> System.Net.WebSockets.WebSocketException (0x80004005): The remote party closed the WebSocket connection without completing the close handshake.
06:01:29    at System.Net.WebSockets.ManagedWebSocket.ThrowIfEOFUnexpected(Boolean throwOnPrematureClosure)
06:01:29    at System.Net.WebSockets.ManagedWebSocket.EnsureBufferContainsAsync(Int32 minimumRequiredBytes, Boolean throwOnPrematureClosure)
06:01:29    at System.Net.WebSockets.ManagedWebSocket.ReceiveAsyncPrivate[TWebSocketReceiveResultGetter,TWebSocketReceiveResult](Memory`1 payloadBuffer, CancellationToken cancellationToken, TWebSocketReceiveResultGetter resultGetter)
06:01:29    at System.Net.WebSockets.ManagedWebSocket.ReceiveAsyncPrivate[TWebSocketReceiveResultGetter,TWebSocketReceiveResult](Memory`1 payloadBuffer, CancellationToken cancellationToken, TWebSocketReceiveResultGetter resultGetter)
06:01:29    at WsProxy.WsProxy.ReadOne(WebSocket socket, CancellationToken token) in /mnt/jenkins/workspace/test-mono-pull-request-wasm/sdks/wasm/Mono.WebAssembly.DebuggerProxy/WsProxy.cs:line 122<---

```

### CODE REVIEW

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build failed

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build failed

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build failed

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build failed

**Comment by kjpou1** [comment]
@monojenkins build failed

**Comment by kjpou1** [comment]
@monojenkins build failed

**Comment by kjpou1** [comment]
@monojenkins build failed

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

File: sdks/wasm/DebuggerTestSuite/Support.cs [hunk 1]
```diff
@@ -61,29 +62,58 @@ async Task OnMessage(string method, JObject args, CancellationToken token)
...
-						}
+
+		    var tries = 0;
+		    using (var readyCTS = new CancellationTokenSource())
+		    {
+			while (!readyCTS.IsCancellationRequested && tries < NUMBER_OF_TRIES)
+			{
+			    if (tries > 0)
+			    {
+				Console.WriteLine($"There seems to have been a problem trying again: Try {tries + 1} of {NUMBER_OF_TRIES}");
```
Comment (kg) [inline sdks/wasm/DebuggerTestSuite/Support.cs]: Minor nit: When I first saw this I thought it meant a retry had failed, not that there was a failure and a retry was occurring.

File: sdks/wasm/Mono.WebAssembly.DebuggerProxy/WsProxy.cs [hunk 2]
```diff
@@ -119,16 +120,23 @@ async Task<string> ReadOne (WebSocket socket, CancellationToken token)
...
-					mem.Write (buff, 0, result.Count);
+					if (result.EndOfMessage) {
+						mem.Write (buff, 0, result.Count);
+						return Encoding.UTF8.GetString (mem.GetBuffer (), 0, (int)mem.Length);
+					} else {
+						mem.Write (buff, 0, result.Count);
+					}
+				}
+				catch (Exception exc) {
+					client_initiated_close.TrySetResult (true);
```
[the remark this answers is not in the document]
Comment (kjpou1) [reply sdks/wasm/Mono.WebAssembly.DebuggerProxy/WsProxy.cs]: I was torn between this as well and went back and forth.  The problem was that I could not recreate locally some of the cases happening on the CI so was imitating some of the errors.  I will take this out.

[the remark this answers is not in the document]
Comment (kg) [reply sdks/wasm/Mono.WebAssembly.DebuggerProxy/WsProxy.cs]: You can leave this as long as you print the exc message to stderr or stdout so it is visible on CI. Would that work?

Alternately, trap only specific exceptions. IIRC there are a couple 'expected' websocket exceptions for when the remote closes, so maybe that's what you want to trap here?

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build failed

**Comment by kjpou1** [comment]
@monojenkins build failed

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by lewing** [comment]
@kjpou1 Where are we with this?

**Comment by lewing** [comment]
@kjpou1 respond or close this please

**Comment by kjpou1** [comment]
Let's keep this open for now.  Still want to look at it.  The problem is that I can not create the exact error locally.

**Comment by lewing** [comment]
I think I was able to reproduce some of the problems easily, in fact I couldn't get a clean run today.  I know devtools can be problematic but I think it is important our proxy is as solid as possible and that our test driver can actually be added to CI.

**Comment by lewing** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
Will take a look.  It will fail locally but not getting the exact error that CI was giving on the websockets connections where a websocket connection was not being found from the devtools interogation.

**Comment by lewing** [comment]
This is very stale now

**Pull request closed without merging**